### ISSUE TITLE
Question about focal loss numeric stablility

### ISSUE THREAD

**Issue by CoinCheung** [open-mmlab/mmdetection#2331]
Hi, 
I am learning the code and I found that when you compute `log(p)` which `p` is `sigmoid(x)`, you use the following method:

https://github.com/open-mmlab/mmdetection/blob/030e2eeab055ede210ae2df6fde31459ad84c7cb/mmdet/ops/sigmoid_focal_loss/src/sigmoid_focal_loss_cuda.cu#L46

I seems to be that you used `log(max(min_float, p))`. However, when you compute `log(1-p)`, you  seem to use a more stable method: 

https://github.com/open-mmlab/mmdetection/blob/030e2eeab055ede210ae2df6fde31459ad84c7cb/mmdet/ops/sigmoid_focal_loss/src/sigmoid_focal_loss_cuda.cu#L49

That when `x>0`, it is `-x - log(1 + exp(-x))` and when `x <0`, it is `-log(1+exp(-x))`. Would you please tell me why you take special care on `log(1-p)`, but use a simple implementation of `log(x)` ? And why did not you use `log(min_float, (1-p))` as does with `log(p)` ?

**Comment by ZwwWayne** [comment]
https://github.com/open-mmlab/mmdetection/issues/1699

**Comment by CoinCheung** [comment]
Yes, I have figured out the reason why we consider differently on the sign of logits when we compute `log(1-p)`, but why do not we do same with `log(p)` ?

**Comment by hellock** [comment]
Just historical reasons. We can both use the stable implementation for `log(p)` and `log(1-p)`.

**Comment by CoinCheung** [comment]
Thanks for replying !!! I am closing this.
